esrally/client/asynchronous.py

```diff
@@ -277,7 +277,12 @@ async def put_lifecycle(self, *args, **kwargs):
             kwargs["name"] = args[0]
 
         if body := kwargs.pop("body", None):
-            kwargs["policy"] = body.get("policy", {})
+            # body may contain a nested 'policy' field, if so, remove it
+            if policy := body.get("policy", {}):
+                kwargs["policy"] = policy
+            else:
+                kwargs["policy"] = body
```

Comment:
I am slightly concerned we assume what the user wants by "fixing" an invalid ILM policy request body by placing the body inside the `policy` element instead of returning a usable error as https://github.com/elastic/rally/issues/1689 requests. However, it is a safe assumption given the create lifecycle policy API only accepts `policy` as the single root body element at the moment. The inconsistency between Rally accepting and handling the invalid request body and Rally/ES documentation could come back to bite us later, and we can cross that bridge when we arrive at it.

In addition to testing this PR using provided schedule, I tested a doubly-invalid `without-policy-field` by replacing `phases` with `phaseses` to see how the typo surfaces. I.e.,

```json
{
  "schedule": [
    {
      "name": "without-policy-field",
      "include-in-reporting": true,
      "operation": {
        "operation-type": "create-ilm-policy",
        "policy-name": "no-policy-field",
        "body": {
          "phaseses": {
            "hot": {
              "min_age": "0ms",
              "actions": {
                "rollover": {
                  "max_primary_shard_size": "50gb",
                  "max_age": "7d"
                },
                "set_priority": {
                  "priority": 100
                }
              }
            }
          }
        }
      }
    },
...
```

We get:

|                        Metric |                 Task |   Value |   Unit |
|------------------------------:|---------------------:|--------:|-------:|
|       Total Young Gen GC time |                      |    0    |      s |
|      Total Young Gen GC count |                      |    0    |        |
|         Total Old Gen GC time |                      |    0    |      s |
|        Total Old Gen GC count |                      |    0    |        |
|   Total Ingest Pipeline count |                      |    0    |        |
|    Total Ingest Pipeline time |                      |    0    |      s |
|  Total Ingest Pipeline failed |                      |    0    |        |
|                    error rate | without-policy-field |  100    |      % |

The problem is very difficult to uncover unless `elastic_transport` debug logging is enabled:

```
2023-11-13 16:16:19,811 -not-actor-/PID:95524 elastic_transport.node DEBUG > PUT /_ilm/policy/no-policy-field HTTP/1.1
> Accept: application/vnd.elasticsearch+json; compatible-with=8
> Connection: keep-alive
> Content-Type: application/vnd.elasticsearch+json; compatible-with=8
> User-Agent: elasticsearch-py/8.6.1 (Python/3.8.13; elastic-transport/8.4.1)
> X-Elastic-Client-Meta: es=8.6.1,py=3.8.13,t=8.4.1,ai=3.8.6
> {"policy":{"phaseses":{"hot":{"min_age":"0ms","actions":{"rollover":{"max_primary_shard_size":"50gb","max_age":"7d"},"set_priority":{"priority":100}}}}}}
< HTTP/1.1 400 Bad Request
< Content-Encoding: gzip
< Content-Length: 203
< Content-Type: application/vnd.elasticsearch+json;compatible-with=8
< X-Elastic-Product: Elasticsearch
< {"error":{"root_cause":[{"type":"x_content_parse_exception","reason":"[1:12] [lifecycle_policy] unknown field [phaseses] did you mean [phases]?"}],"type":"x_content_parse_exception","reason":"[1:23] [put_lifecycle_request] failed to parse field [policy]","caused_by":{"type":"x_content_parse_exception","reason":"[1:12] [lifecycle_policy] unknown field [phaseses] did you mean [phases]?"}},"status":400}
2023-11-13 16:16:19,811 -not-actor-/PID:95524 elastic_transport.transport INFO PUT http://127.0.0.1:39200/_ilm/policy/no-policy-field [status:400 duration:0.017s]
2023-11-13 16:16:19,812 ActorAddr-(T|:51435)/PID:95524 esrally.driver.driver INFO Worker[0] finished executing tasks ['without-policy-field'] in 0.019253 seconds
```

What would you think about leaving the missing `policy` element as invalid, not having Rally validate it, and just returning the 400 error description returned by Elasticsearch?

Reply:
> The inconsistency between Rally accepting and handling the invalid request body and Rally/ES documentation could come back to bite us later

I agree here, for some reason I thought you could submit an ILM policy without the root element, but that's not correct.

> The problem is very difficult to uncover unless elastic_transport debug logging is enabled:

This is somewhat orthogonal to this PR, as the burying of the error message is related to this https://github.com/elastic/rally/issues/1581 - if you run your test with `--on-error=abort` you'll get:
```
Running with-policy-field                                                      [  0% done]
[ERROR] Cannot race. Error in load generator [0]
        Cannot run task [with-policy-field]: Request returned an error. Error type: api, Description: x_content_parse_exception ({'error': {'root_cause': [{'type': 'x_content_parse_exception', 'reason': '[1:12] [lifecycle_policy] unknown field [phassses] did you mean [phases]?'}], 'type': 'x_content_parse_exception', 'reason': '[1:23] [put_lifecycle_request] failed to parse field [policy]', 'caused_by': {'type': 'x_content_parse_exception', 'reason': '[1:12] [lifecycle_policy] unknown field [phassses] did you mean [phases]?'}}, 'status': 400})
```

I think what I'll do here is fix the misleading error message reported in https://github.com/elastic/rally/issues/1689, but we won't handle request bodies missing the root `policy` element.

Reply:
So turns out we don't want ES to return the exception in this case anyway, because the client adds the root `policy` field we end up with an even more misleading error:
```
[without-policy-field]: Request returned an error. Error type: api, Description: x_content_parse_exception ({'error': {'root_cause': [{'type': 'x_content_parse_exception', 'reason': '[1:12] [put_lifecycle_request] failed to parse field [policy]'}], 'type': 'x_content_parse_exception', 'reason': '[1:12] [put_lifecycle_request] failed to parse field [policy]', 'caused_by': {'type': 'illegal_argument_exception', 'reason': 'Required [phases]'}}, 'status': 400})
```

Reply:
I pushed https://github.com/elastic/rally/pull/1801/commits/8d894181d44e50ad790eac84e88445d76195284c, which behaves like so (even without `--on-error=abort`):
```
$ esrally race --distribution-version=8.11.0 --track-path=simple-ilm.json --kill-running-processes 

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [7f7f9d10-f04a-48f7-9c31-8343d693f861]
[INFO] Preparing for race ...
[INFO] Racing on track [simple-ilm] and car ['defaults'] with version [8.11.0].

Running without-policy-field                                                   [  0% done]
[ERROR] Cannot race. Error in load generator [0]
        Cannot run task [without-policy-field]: Request body does not contain the expected root field [policy]. Please ensure that the request body contains a top-level 'policy' field and try again.
```